# Project Euler
________________

# Problem 1: Multiples of 3 or 5

If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.

Find the sum of all the multiples of 3 or 5 below 1000.

In [1]:
import numpy as np

In [20]:
arr = np.arange(999) + 1
arr

array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
        40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
        53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
        66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
        79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
        92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103, 104,
       105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117,
       118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130,
       131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143,
       144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156,
       157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169,
       170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 18

In [22]:
bool_arr = np.logical_or(arr % 3 == 0, arr % 5 == 0)
bool_arr

array([False, False,  True, False,  True,  True, False, False,  True,
        True, False,  True, False, False,  True, False, False,  True,
       False,  True,  True, False, False,  True,  True, False,  True,
       False, False,  True, False, False,  True, False,  True,  True,
       False, False,  True,  True, False,  True, False, False,  True,
       False, False,  True, False,  True,  True, False, False,  True,
        True, False,  True, False, False,  True, False, False,  True,
       False,  True,  True, False, False,  True,  True, False,  True,
       False, False,  True, False, False,  True, False,  True,  True,
       False, False,  True,  True, False,  True, False, False,  True,
       False, False,  True, False,  True,  True, False, False,  True,
        True, False,  True, False, False,  True, False, False,  True,
       False,  True,  True, False, False,  True,  True, False,  True,
       False, False,  True, False, False,  True, False,  True,  True,
       False, False,

In [23]:
bool_arr @ arr

233168

## **Answer:** 
233168

### Alternative:

There's a repeating pattern every 15 numbers:

In [25]:
bool_arr[:15], bool_arr[15:15+15]

(array([False, False,  True, False,  True,  True, False, False,  True,
         True, False,  True, False, False,  True]),
 array([False, False,  True, False,  True,  True, False, False,  True,
         True, False,  True, False, False,  True]))

In [27]:
all(bool_arr[:15] == bool_arr[15:15+15])

True

In [28]:
bool_arr[:15] @ arr[:15]

60

In [31]:
def integer_division(n, d):
    return n // d , n % d

In [32]:
integer_division(999, 15)

(66, 9)

In [35]:
def triangular_number(n):
    return (n * (n-1)) // 2

In [36]:
60 * triangular_number(66) * 15 + bool_arr[-9:] @ arr[-9:]

1934483

# Problem 2: Even Fibonacci numbers

Each new term in the Fibonacci sequence is generated by adding the previous two terms. By starting with 1 and 2, the first 10 terms will be:

> 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, ...

By considering the terms in the Fibonacci sequence whose values do not exceed four million, find the sum of the even-valued terms.

In [39]:
import functools

In [45]:
@functools.lru_cache(maxsize=5)
def fibonacci(n):
    # Returns the n-th fibonacci number.
    if n < 1:
        raise ValueError("Fibonacci is a sequence. No negative indexes allowed.")
    if n == 1:
        return 1
    if n == 2:
        return 2
    return fibonacci(n-2) + fibonacci(n-1)

for i in range(1, 10):
    print(fibonacci(i))

1
2
3
5
8
13
21
34
55


In [56]:
fib_arr = np.array(list(map(fibonacci, np.arange(1, 50+1))))
fib_arr

array([          1,           2,           3,           5,           8,
                13,          21,          34,          55,          89,
               144,         233,         377,         610,         987,
              1597,        2584,        4181,        6765,       10946,
             17711,       28657,       46368,       75025,      121393,
            196418,      317811,      514229,      832040,     1346269,
           2178309,     3524578,     5702887,     9227465,    14930352,
          24157817,    39088169,    63245986,   102334155,   165580141,
         267914296,   433494437,   701408733,  1134903170,  1836311903,
        2971215073,  4807526976,  7778742049, 12586269025, 20365011074])

In [62]:
bool_arr = np.logical_and(fib_arr < 4E6, fib_arr % 2 == 0)
bool_arr

array([False,  True, False, False,  True, False, False,  True, False,
       False,  True, False, False,  True, False, False,  True, False,
       False,  True, False, False,  True, False, False,  True, False,
       False,  True, False, False,  True, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False])

In [63]:
bool_arr @ fib_arr

4613732

### Answer:
4613732

# Problem 3: Largest prime factor

The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600851475143 ?

In [69]:
from numba import njit

In [76]:
600851475143 ** 0.5

775146.0992245268

In [482]:
@njit
def prime_search(n):
    """Returns a numpy array of primes up to n."""
    sieve = np.arange(3, n+1, 2)
    len_sieve = len(sieve)
    primes = np.zeros(len_sieve)
    primes[0] = 2
    j = 1
    # limit = (len_sieve ** 0.5) // 2
    limit = len_sieve
    for i in range(len_sieve):
        curr_numb = sieve[i]
        if curr_numb != 0:
            if i < limit:
                sieve[i::curr_numb] *= 0
            primes[j] = curr_numb
            j += 1           
    return primes[primes!=0]

In [483]:
prime_search(100)

array([ 2.,  3.,  5.,  7., 11., 13., 17., 19., 23., 29., 31., 37., 41.,
       43., 47., 53., 59., 61., 67., 71., 73., 79., 83., 89., 97.])

In [247]:
@njit
def factor(n):
    limit = int(n ** 0.5 * 2)
    primes = prime_search(limit)
    factors = np.zeros(limit)
    j = 0
    for p in primes:
        while n % p == 0:
            factors[j] = p
            j += 1
            n /= p
    return factors[factors!=0]

@njit
def factor_v2(n):
    limit = int(n ** 0.5 * 2)
    primes = prime_search(limit)
    
    factors = primes[n % primes == 0]

    return factors

In [250]:
%timeit factor(71*73*34)
%timeit factor_v2(71*73*34)

24.8 µs ± 2.43 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
32.2 µs ± 2.57 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [251]:
%timeit factor(600851475143)
%timeit factor_v2(600851475143)

67.3 ms ± 3.03 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
71.6 ms ± 7.99 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [253]:
factor_v2(600851475143)

array([  71.,  839., 1471., 6857.])

### Answer:
6857

# Problem 4: Largest palindrome product

A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.

In [279]:
import math

In [278]:
powers = 10 ** np.arange(6, -1, -1); print(powers)
(123456 % powers[:-1]) // powers[1:]

[1000000  100000   10000    1000     100      10       1]


array([1, 2, 3, 4, 5, 6])

In [417]:
def split_number(n:int) -> np.array:
    """
    123456 -> np.array([1, 2, 3, 4, 5, 6])
    """
    if n < 0:
        raise ValueError("Can't split negative numbers.")
    if n != int(n):
        raise TypeError("Can't split floats.")
        
    if n in [0, 1]:
        return np.array([n])
    
    # number_of_digits = math.ceil(math.log(n, 10)) # <-- probably good to remove for faster speed
    number_of_digits = len(str(n))
    powers = 10 ** np.arange(number_of_digits, -1, -1) # (10^n, 10^(n-1), 10^(n-2), ..., 10, 1)
    # % cuts the front part, // cuts the end part. That leaves only the digit we want in each position
    return (n % powers[:-1]) // powers[1:]

def is_palindrome(n:int) -> bool:
    arr = split_number(n)
    # if it's the same reversed as it is normally, it's palindrome.
    return all(arr == arr[::-1])

In [418]:
split_number(999*999)

array([9, 9, 8, 0, 0, 1])

In [419]:
is_palindrome(9009)

True

In [423]:
%timeit
result = 0

for i in range(999, 0, -1):
    if i * i < result:
        break
    for j in range(i, 0, -1):
        if is_palindrome(prod := i*j):
            # print(f"{i}*{j} = {prod}")
            if result < prod:
                result = prod
            break

print(result)

906609


### Answer:
906.609

# Problem 5: Smallest multiple

2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?

In [394]:
n = 10
print(f"n = {n}")
primes_under_n = prime_search(n)
print(primes_under_n)
powers = np.log(n) // np.log(primes_under_n)
print(powers)
max_pure_powers = primes_under_n ** powers
print(max_pure_powers)

print("===")
n = 20
print(f"n = {n}")
primes_under_n = prime_search(n)
print(primes_under_n)
powers = np.log(n) // np.log(primes_under_n)
print(powers)
max_pure_powers = primes_under_n ** powers
print(max_pure_powers)

n = 10
[2. 3. 5. 7.]
[3. 2. 1. 1.]
[8. 9. 5. 7.]
===
n = 20
[ 2.  3.  5.  7. 11. 13. 17. 19.]
[4. 2. 1. 1. 1. 1. 1. 1.]
[16.  9.  5.  7. 11. 13. 17. 19.]


In [389]:
def smallest_multiple_of_all_under_n(n):
    primes_under_n = prime_search(n)
    powers = np.log(n) // np.log(primes)
    max_pure_powers = primes ** powers
    return max_pure_powers.prod()

In [390]:
smallest_multiple_of_all_under_n(20)

232792560.0

### Answer:
232.792.560

# Problem 6: Sum square difference

The sum of the squares of the first ten natural numbers is,

$$1^2 + 2^2+ \dots + 10^2 = 385$$

The square of the sum of the first ten natural numbers is,

$$(1 + 2+ \dots + 10)^2 = 3025$$

Hence the difference between the sum of the squares of the first ten natural numbers and the square of the sum is $ 3025 - 385 = 2640$ .

Find the difference between the sum of the squares of the first one hundred natural numbers and the square of the sum.

In [396]:
first_100_numbers = np.arange(1, 100+1)
first_100_numbers

array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
        40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
        53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
        66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
        79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
        92,  93,  94,  95,  96,  97,  98,  99, 100])

In [399]:
first_100_numbers ** 2

array([    1,     4,     9,    16,    25,    36,    49,    64,    81,
         100,   121,   144,   169,   196,   225,   256,   289,   324,
         361,   400,   441,   484,   529,   576,   625,   676,   729,
         784,   841,   900,   961,  1024,  1089,  1156,  1225,  1296,
        1369,  1444,  1521,  1600,  1681,  1764,  1849,  1936,  2025,
        2116,  2209,  2304,  2401,  2500,  2601,  2704,  2809,  2916,
        3025,  3136,  3249,  3364,  3481,  3600,  3721,  3844,  3969,
        4096,  4225,  4356,  4489,  4624,  4761,  4900,  5041,  5184,
        5329,  5476,  5625,  5776,  5929,  6084,  6241,  6400,  6561,
        6724,  6889,  7056,  7225,  7396,  7569,  7744,  7921,  8100,
        8281,  8464,  8649,  8836,  9025,  9216,  9409,  9604,  9801,
       10000])

In [401]:
first_100_numbers.sum() ** 2 - (first_100_numbers ** 2).sum()

25164150

### Answer: 
25.164.150

# Problem 7: 10001st prime

By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.

What is the 10 001st prime number?

In [404]:
len(prime_search(1_000_000))

80515

In [406]:
prime_search(1_000_000)[5]

13.0

In [407]:
prime_search(1_000_000)[10_000]

104743.0

# Answer: 

104.743

# Problem 8: Largest product in a series

The four adjacent digits in the 1000-digit number that have the greatest product are 9 × 9 × 8 × 9 = 5832.

73167176531330624919225119674426574742355349194934  
96983520312774506326239578318016984801869478851843  
85861560789112949495459501737958331952853208805511  
12540698747158523863050715693290963295227443043557  
66896648950445244523161731856403098711121722383113  
62229893423380308135336276614282806444486645238749  
30358907296290491560440772390713810515859307960866  
70172427121883998797908792274921901699720888093776  
65727333001053367881220235421809751254540594752243  
52584907711670556013604839586446706324415722155397  
53697817977846174064955149290862569321978468622482  
83972241375657056057490261407972968652414535100474  
82166370484403199890008895243450658541227588666881  
16427171479924442928230863465674813919123162824586  
17866458359124566529476545682848912883142607690042  
24219022671055626321111109370544217506941658960408  
07198403850962455444362981230987879927244284909188  
84580156166097919133875499200524063689912560717606  
05886116467109405077541002256983155200055935729725  
71636269561882670428252483600823257530420752963450  

Find the thirteen adjacent digits in the 1000-digit number that have the greatest product. What is the value of this product?

In [408]:
long_number_str = "73167176531330624919225119674426574742355349194934\
96983520312774506326239578318016984801869478851843\
85861560789112949495459501737958331952853208805511\
12540698747158523863050715693290963295227443043557\
66896648950445244523161731856403098711121722383113\
62229893423380308135336276614282806444486645238749\
30358907296290491560440772390713810515859307960866\
70172427121883998797908792274921901699720888093776\
65727333001053367881220235421809751254540594752243\
52584907711670556013604839586446706324415722155397\
53697817977846174064955149290862569321978468622482\
83972241375657056057490261407972968652414535100474\
82166370484403199890008895243450658541227588666881\
16427171479924442928230863465674813919123162824586\
17866458359124566529476545682848912883142607690042\
24219022671055626321111109370544217506941658960408\
07198403850962455444362981230987879927244284909188\
84580156166097919133875499200524063689912560717606\
05886116467109405077541002256983155200055935729725\
71636269561882670428252483600823257530420752963450"

In [459]:
digits = np.array([int(digit) for digit in long_number_str])
digits

array([7, 3, 1, 6, 7, 1, 7, 6, 5, 3, 1, 3, 3, 0, 6, 2, 4, 9, 1, 9, 2, 2,
       5, 1, 1, 9, 6, 7, 4, 4, 2, 6, 5, 7, 4, 7, 4, 2, 3, 5, 5, 3, 4, 9,
       1, 9, 4, 9, 3, 4, 9, 6, 9, 8, 3, 5, 2, 0, 3, 1, 2, 7, 7, 4, 5, 0,
       6, 3, 2, 6, 2, 3, 9, 5, 7, 8, 3, 1, 8, 0, 1, 6, 9, 8, 4, 8, 0, 1,
       8, 6, 9, 4, 7, 8, 8, 5, 1, 8, 4, 3, 8, 5, 8, 6, 1, 5, 6, 0, 7, 8,
       9, 1, 1, 2, 9, 4, 9, 4, 9, 5, 4, 5, 9, 5, 0, 1, 7, 3, 7, 9, 5, 8,
       3, 3, 1, 9, 5, 2, 8, 5, 3, 2, 0, 8, 8, 0, 5, 5, 1, 1, 1, 2, 5, 4,
       0, 6, 9, 8, 7, 4, 7, 1, 5, 8, 5, 2, 3, 8, 6, 3, 0, 5, 0, 7, 1, 5,
       6, 9, 3, 2, 9, 0, 9, 6, 3, 2, 9, 5, 2, 2, 7, 4, 4, 3, 0, 4, 3, 5,
       5, 7, 6, 6, 8, 9, 6, 6, 4, 8, 9, 5, 0, 4, 4, 5, 2, 4, 4, 5, 2, 3,
       1, 6, 1, 7, 3, 1, 8, 5, 6, 4, 0, 3, 0, 9, 8, 7, 1, 1, 1, 2, 1, 7,
       2, 2, 3, 8, 3, 1, 1, 3, 6, 2, 2, 2, 9, 8, 9, 3, 4, 2, 3, 3, 8, 0,
       3, 0, 8, 1, 3, 5, 3, 3, 6, 2, 7, 6, 6, 1, 4, 2, 8, 2, 8, 0, 6, 4,
       4, 4, 4, 8, 6, 6, 4, 5, 2, 3, 8, 7, 4, 9, 3,

In [454]:
from numpy.lib.stride_tricks import sliding_window_view

In [461]:
sliding_window_view(digits, window_shape = 4).prod(axis=1).max()

5832

In [462]:
sliding_window_view(digits, window_shape = 13).prod(axis=1).max()

23514624000

### Answer: 
23514624000

# Problem 9: Special Pythagorean triplet

A Pythagorean triplet is a set of three natural numbers, $a < b < c$, for which,

$$a^2 + b^2 = c^2$$
For example, $3^2 + 4^2 = 9 + 16 = 25 = 5^2$.

There exists exactly one Pythagorean triplet for which $a + b + c = 1000$.  
Find the product $abc$.

In [472]:
from numba import prange

In [477]:
@njit
def find_special_pythagorean(n):
    for a in range(1, n):
        for b in range(a+1, n-a):
            c = (a ** 2 + b ** 2) ** 0.5
            if a + b + c ==  n:
                return a * b * c

In [478]:
find_special_pythagorean(1000)

31875000.0

### Answer:
31.875.000

# Problem 10: Summation of primes

The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.

Find the sum of all the primes below two million.

In [486]:
prime_search(10)

array([2., 3., 5., 7.])

In [485]:
prime_search(10).sum()

17.0

In [484]:
prime_search(2_000_000).sum()

142913828922.0

### Answer:
142.913.828.922